crates/provider/src/heart.rs

```diff
@@ -499,15 +619,43 @@ impl<S: Stream<Item = Block> + Unpin + 'static> Heartbeat<S> {
                 select! {
                     biased;
 
-                    // Watch for new transactions.
+                    // Indicates that something went wrong, likely client was dropped.
+                    _ = shutdown.recv() => {
+                        break 'shutdown;
+                    },
+
+                    // Watch for new tx watching requests.
                     ix_opt = ixns.recv() => match ix_opt {
-                        Some(to_watch) => self.handle_watch_ix(to_watch),
+                        Some(to_watch) => self.handle_watch_ix(to_watch, &mut receipt_tasks),
                         None => break 'shutdown, // ix channel is closed
                     },
 
+                    // Handle potential receipts for newly requested transaction.
+                    Some(Some(receipt)) = receipt_tasks.fetch_receipt_tasks.next() => {
```

Comment:
[should be flattened?](https://doc.rust-lang.org/std/option/enum.Option.html#method.flatten)